datafusion/expr/src/logical_plan/builder.rs

```diff
@@ -502,6 +505,25 @@ impl LogicalPlanBuilder {
             ));
         }
 
+        let filter = if let Some(expr) = filter {
+            // ambiguous check
+            ensure_any_column_reference_is_unambiguous(
+                &expr,
+                &[self.schema(), right.schema()],
+            )?;
+
+            // normalize all columns in expression
+            let using_columns = expr.to_columns()?;
+            let filter = normalize_col_with_schemas(
+                expr,
+                &[self.schema(), right.schema()],
+                &[using_columns],
+            )?;
+            Some(filter)
+        } else {
+            None
+        };
+
```

Comment:
related to https://github.com/apache/arrow-datafusion/issues/4196

fix bug where you could do dataframe join with ambiguous column for the filter expr

instead of having the check done in both DataFrame join api and SQL planner join mod, unify by having check done inside the logical plan builder

this is technically an unrelated fix to the actual issue, so i can extract into separate issue if needed

Reply:
I think it is fine to include in this PR as long as it also has a test (for ambiguity check using the DataFrame API)

Reply:
test added